code-gen/parse-csl/src/Csl/Gen/Utils.hs

```diff
@@ -0,0 +1,230 @@
+  withSemicolon,
+  wrapText,
+) where
+
+import Control.Monad (guard)
+import Csl.Types
+import Data.Char (isUpper, toLower)
+import Data.List qualified as L (foldl', null)
+import Data.List.Extra (dropPrefix, isPrefixOf)
+import Data.List.Extra qualified as L (
+  nub,
+  sort,
+  stripSuffix,
+ )
+import Data.Maybe (
+  fromMaybe,
+  listToMaybe,
+  mapMaybe,
+ )
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Text (Text)
+import Data.Text qualified as T (pack, replace, unpack)
+import Data.Text.Manipulate qualified as T (lowerHead, toCamel, toTitle, upperHead)
+
+toName :: String -> String
+toName = lowerHead . substFirst . subst . toCamel
+
+toTypePrefix :: String -> String
+toTypePrefix = lowerHead . subst . upperHead . toCamel
+
+toType :: String -> String
+toType = dropPrefix "ValuesTypeof" . subst . upperHead . toCamel
+
+isCommon :: Fun -> Bool
+isCommon (Fun "free" _ _) = True
+isCommon (Fun "to_bytes" _ _) = True
+isCommon (Fun "from_bytes" _ _) = True
+isCommon (Fun "to_hex" _ _) = True
+isCommon (Fun "from_hex" _ _) = True
+isCommon (Fun "to_json" _ _) = True
+isCommon (Fun "from_json" _ _) = True
+isCommon (Fun "to_js_value" _ _) = True
+isCommon (Fun "from_js_value" _ _) = True
+-- sometimes these are with prefixes, sometimes not. they resist abstraction
+-- isCommon (Fun "from_bech32" _ _) = True
+-- isCommon (Fun "to_bech32" _ _) = True
+isCommon (Fun "len" _ _) = True
+isCommon (Fun "add" _ _) = True
+isCommon (Fun "insert" _ _) = True
+isCommon (Fun "get" _ _) = True
+isCommon (Fun "keys" _ _) = True
+isCommon _ = False
+
+withSemicolon :: String -> String
+withSemicolon = flip mappend ";"
+
+replaceFirst :: String -> String -> String -> String
+replaceFirst from to str
+  | from `isPrefixOf` str = to <> drop (length from) str
+  | otherwise = str
+
+substFirst :: String -> String
+substFirst = replacesBy replaceFirst []
+
+subst :: String -> String
+subst =
+  replacesBy
+    replace
+    [ ("Uint8Array", "ByteArray")
+    , ("Void", "Unit")
+    , ("JSON", "Json")
+    ]
+
+replacesBy :: (String -> String -> String -> String) -> [(String, String)] -> String -> String
+replacesBy repl = L.foldl' (\res a -> res . uncurry repl a) id
+
+replace :: String -> String -> String -> String
+replace from to = T.unpack . T.replace (T.pack from) (T.pack to) . T.pack
+
+wrapText :: (Text -> Text) -> (String -> String)
+wrapText f = T.unpack . f . T.pack
+
+toCamel :: String -> String
+toCamel = wrapText T.toCamel
+
+upperHead :: String -> String
+upperHead = wrapText T.upperHead
+
+lowerHead :: String -> String
+lowerHead str
+  | L.null post = fmap toLower str
+  | otherwise =
+      case pre of
+        [] -> wrapText T.lowerHead post
+        [a] -> toLower a : post
+        _ -> map toLower (init pre) <> [last pre] <> post
+  where
+    (pre, post) = span isUpper str
+
+filterMethods :: Class -> [Method]
+filterMethods (Class name ms)
+  -- CostModel is a special case: it looks like a mapping type, but isn't
```

Comment:
This comment should be fixed to also include `VotingProcedures`, which is not a normal map and should have its methods exposed.

Reply:
Addressed in 09b97f4.